django/core/management/commands/shell.py

```diff
@@ -111,6 +127,55 @@ def python(self, options):
         # Start the interactive interpreter.
         code.interact(local=imported_objects)
 
+    def get_app_name(self, app_module):
+        rparts = list(reversed(app_module.split(".")))
+        try:
+            try:
+                return rparts[rparts.index(MODELS_MODULE_NAME) + 1]
+            except ValueError:
+                return rparts[1]
+        except IndexError:
+            return app_module
+
+    def get_apps_and_models(self):
+        for app_conf in apps.get_app_configs():
+            if app_conf.models_module:
+                yield app_conf.models_module, app_conf.get_models()
+
+    def get_namespace(self, options, style):
+        """Override this method with import_items in order to import extra_things"""
+
+        imported_objects = {}
+
+        for app_module, app_models in self.get_apps_and_models():
+            if not app_models:
+                continue
+            model_labels = []
+            for model in app_models:
+                if model.__module__:
+                    try:
+                        imported_objects[model.__name__] = import_string(
+                            "%s.%s" % (model.__module__, model.__name__)
+                        )
```

Comment:
I think you’ve misunderstood what `model` and `import_string()` are here. This `import_string()` invocation will always just return the model class, which is what `model` is to begin with. For example, using a project I have open:

```
In [1]: from django.utils.module_loading import import_string

In [2]: from example.core.models import User

In [3]: model = User

In [4]: import_string("%s.%s" % (model.__module__, model.__name__))
Out[4]: example.core.models.User

In [5]: import_string("%s.%s" % (model.__module__, model.__name__)) is User
Out[5]: True
```

So, you can use `model` directly:

```suggestion
                        imported_objects[model.__name__] = model
```


Also, in terms of overriding, this implementation gives precedence to later apps for duplicate model names. We need to give precedence to *earlier* apps.